### ISSUE TITLE
How can I implement parameters of a Model class which could change over batches/epochs during training

### ISSUE THREAD

**Issue by MeiqiGuo** [allenai/allennlp#4248]
When I implement my class of Model, how can I add a parameter which could change over batches/epochs (somewhat like learning rate scheduler)? For example, I would like to decrease the "p" of a dropout layer in my Model after every batch during the training. Is there a way to do it please? Thanks in advance!

**Comment by dirkgr** [comment]
You could use a `BatchCallback`: https://github.com/allenai/allennlp/blob/master/allennlp/training/trainer.py#L102

The callback has access to the trainer, which has access to the model. So you could set some parameter on the model that way after every batch.

**Comment by MeiqiGuo** [comment]
@dirkgr Thank you so much, this is very helpful! 
Do you know any code example which uses it for some similar purpose? Thanks in advance!

**Comment by dirkgr** [comment]
We don't have great examples for this, but here is a test that uses `BatchCallback` end-to-end:

https://github.com/allenai/allennlp/blob/f27475aaeba87052e277df83beaadf002f4fd9da/tests/commands/train_test.py#L247

There is a model config in the test that you can adopt in your own code. The callback itself is here: https://github.com/allenai/allennlp/blob/f27475aaeba87052e277df83beaadf002f4fd9da/tests/commands/train_test.py#L33

This callback only writes a few log messages, but you can do anything in there, including set a parameter on the model, which you can then read in your `Model.forward()` method.

**Comment by MeiqiGuo** [comment]
@dirkgr It seems unclear for me how to use BatchCallBack to feed parameters to Model.forward(). I understand how to use it to do some loggings, however, it is different from intreating with other classes (such as Model). Would you please clarify it? Thanks!

**Comment by matt-gardner** [comment]
In the `BatchCallback` (and `EpochCallback`), you can set `trainer.model.batch_num = counter` or `trainer.model.epoch_num = counter`, and then read that variable that you set inside `model.forward()`.

**Comment by matt-gardner** [comment]
This may not be a safe operation in a distributed setting; I'm not sure how this would interact.  Or maybe each thread would write to its own thread-local objects, so it all works out...?  🤷‍♂️ Not sure exactly how the multi-threading works here.

**Comment by MeiqiGuo** [comment]
I see. Thank you Matt @matt-gardner for your message!
